src/ray/object_manager/common.cc

```diff
@@ -139,19 +164,25 @@ Status PlasmaObjectHeader::WriteRelease(Semaphores &sem) {
   return Status::OK();
 }
 
-Status PlasmaObjectHeader::ReadAcquire(Semaphores &sem,
-                                       int64_t version_to_read,
-                                       int64_t &version_read) {
+Status PlasmaObjectHeader::ReadAcquire(
+    Semaphores &sem,
+    int64_t version_to_read,
+    int64_t &version_read,
+    const std::unique_ptr<std::chrono::steady_clock::time_point> &timeout_point) {
   RAY_CHECK(sem.header_sem);
 
-  RAY_RETURN_NOT_OK(TryToAcquireSemaphore(sem.header_sem));
+  RAY_RETURN_NOT_OK(TryToAcquireSemaphore(sem.header_sem, timeout_point));
 
   // TODO(jhumphri): Wouldn't a futex be better here than polling?
   // Wait for the requested version (or a more recent one) to be sealed.
   while (version < version_to_read || !is_sealed) {
     RAY_CHECK_EQ(sem_post(sem.header_sem), 0);
     sched_yield();
-    RAY_RETURN_NOT_OK(TryToAcquireSemaphore(sem.header_sem));
+    if (timeout_point != nullptr && std::chrono::steady_clock::now() >= *timeout_point) {
```

Comment:
Is this check necessary? Won't `TryToAcquireSemaphore()` return an error status if the timeout has expired?

Reply:
Yes, we need it to make sure we get the desired version before timeout. Added a comment to clarify.